// @cosmjs/ledger-amino/build/demo/web.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const amino_1 = require("@cosmjs/amino");
const crypto_1 = require("@cosmjs/crypto");
const encoding_1 = require("@cosmjs/encoding");
const math_1 = require("@cosmjs/math");
const utils_1 = require("@cosmjs/utils");
const hw_transport_webusb_1 = __importDefault(require("@ledgerhq/hw-transport-webusb"));
const ledgersigner_1 = require("../ledgersigner");
const accountNumbers = [0, 1, 2, 10];
const paths = accountNumbers.map(amino_1.makeCosmoshubPath);
let accounts = [];
function createSignDoc(accountNumber, address) {
    const signDoc = {
        // eslint-disable-next-line @typescript-eslint/naming-convention
        chain_id: "testing",
        // eslint-disable-next-line @typescript-eslint/naming-convention
        account_number: `${accountNumber}`,
        sequence: "0",
        fee: {
            amount: [{ amount: "100", denom: "ucosm" }],
            gas: "250",
        },
        memo: "Some memo",
        msgs: [
            {
                type: "cosmos-sdk/MsgSend",
                value: {
                    amount: [
                        {
                            amount: "1234567",
                            denom: "ucosm",
                        },
                    ],
                    // eslint-disable-next-line @typescript-eslint/naming-convention
                    from_address: address,
                    // eslint-disable-next-line @typescript-eslint/naming-convention
                    to_address: address,
                },
            },
        ],
    };
    return JSON.stringify(signDoc, null, 2);
}
window.updateMessage = (accountNumberInput) => {
    (0, utils_1.assert)(typeof accountNumberInput === "string");
    const accountNumber = math_1.Uint53.fromString(accountNumberInput).toNumber();
    const account = accounts[accountNumber];
    if (account === undefined) {
        return;
    }
    const address = accounts[accountNumber].address;
    const addressInput = document.getElementById("address");
    addressInput.value = address;
    const signDocTextArea = document.getElementById("sign-doc");
    signDocTextArea.textContent = createSignDoc(accountNumber, address);
};
window.setPath = (accountNumberInput) => {
    (0, utils_1.assert)(typeof accountNumberInput === "string");
    const accountNumber = math_1.Uint53.fromString(accountNumberInput).toNumber();
    const path = (0, crypto_1.pathToString)(paths[accountNumber]);
    const pathInput = document.getElementById("path");
    pathInput.value = path;
};
// This must be called by the user an cannot be done on load (see "TransportWebUSBGestureRequired").
window.createSigner = async function createSigner() {
    const interactiveTimeout = 120000;
    const ledgerTransport = await hw_transport_webusb_1.default.create(interactiveTimeout, interactiveTimeout);
    return new ledgersigner_1.LedgerSigner(ledgerTransport, {
        testModeAllowed: true,
        hdPaths: paths,
    });
};
window.getAccounts = async function getAccounts(signer) {
    if (signer === undefined) {
        console.error("Please wait for transport to connect");
        return;
    }
    const accountNumberInput1 = document.getElementById("account-number1");
    const accountNumberInput2 = document.getElementById("account-number2");
    const addressInput = document.getElementById("address");
    const accountsDiv = document.getElementById("accounts");
    const signDocTextArea = document.getElementById("sign-doc");
    accountsDiv.textContent = "Loading...";
    try {
        accounts = await signer.getAccounts();
        const prettyAccounts = accounts.map((account) => ({
            ...account,
            pubkey: (0, encoding_1.toBase64)(account.pubkey),
        }));
        accountsDiv.textContent = JSON.stringify(prettyAccounts, null, "\n");
        const accountNumber = 0;
        // Show address block
        accountNumberInput1.max = accounts.length - 1;
        accountNumberInput1.value = accountNumber;
        window.setPath(accountNumber.toString());
        // Sign block
        accountNumberInput2.max = accounts.length - 1;
        accountNumberInput2.value = accountNumber;
        const address = accounts[0].address;
        addressInput.value = address;
        signDocTextArea.textContent = createSignDoc(accountNumber, address);
    }
    catch (error) {
        console.error(error);
        accountsDiv.textContent = error;
    }
};
window.showAddress = async function showAddress(signer) {
    if (signer === undefined) {
        console.error("Please wait for transport to connect");
        return;
    }
    const path = (0, crypto_1.stringToPath)(document.getElementById("path").value);
    await signer.showAddress(path);
};
window.sign = async function sign(signer) {
    if (signer === undefined) {
        console.error("Please wait for transport to connect");
        return;
    }
    const signatureDiv = document.getElementById("signature");
    signatureDiv.textContent = "Loading...";
    try {
        const address = document.getElementById("address").value;
        const signDocJson = document.getElementById("sign-doc").textContent;
        const signDoc = JSON.parse(signDocJson);
        const signature = await signer.signAmino(address, signDoc);
        signatureDiv.textContent = JSON.stringify(signature, null, "\t");
    }
    catch (error) {
        signatureDiv.textContent = error;
    }
};
