// @cosmjs/ledger-amino/build/testutils.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.simapp = exports.pendingWithoutSimapp = exports.simappEnabled = exports.pendingWithoutLedger = exports.ledgerEnabled = exports.faucet = void 0;
exports.faucet = {
    mnemonic: "economy stock theory fatal elder harbor betray wasp final emotion task crumble siren bottom lizard educate guess current outdoor pair theory focus wife stone",
    pubkey: {
        type: "tendermint/PubKeySecp256k1",
        value: "A08EGB7ro1ORuFhjOnZcSgwYlpe0DSFjVNUIkNNQxwKQ",
    },
    address: "cosmos1pkptre7fdkl6gfrzlesjjvhxhlc3r4gmmk8rs6",
};
function ledgerEnabled() {
    return !!process.env.LEDGER_ENABLED;
}
exports.ledgerEnabled = ledgerEnabled;
function pendingWithoutLedger() {
    if (!ledgerEnabled()) {
        return pending("Set LEDGER_ENABLED to enable Ledger-based tests");
    }
}
exports.pendingWithoutLedger = pendingWithoutLedger;
function simappEnabled() {
    return !!process.env.SIMAPP42_ENABLED || !!process.env.SIMAPP44_ENABLED;
}
exports.simappEnabled = simappEnabled;
function pendingWithoutSimapp() {
    if (!simappEnabled()) {
        return pending("Set SIMAPP42_ENABLED or SIMAPP44_ENABLED to enable Simapp-based tests");
    }
}
exports.pendingWithoutSimapp = pendingWithoutSimapp;
exports.simapp = {
    endpoint: "ws://localhost:26658",
    chainId: "simd-testing",
};
